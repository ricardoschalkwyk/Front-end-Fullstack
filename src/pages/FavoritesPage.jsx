import { HeartIcon } from "@heroicons/react/outline";
import React, { useEffect, useState } from "react";

import PropTypes from "prop-types";

import Api from "../Api";
import FavoriteItem from "../components/FavoriteItem";

// mock prop is passed for testing
function FavoritesPage({ mock = [] }) {
  // This state holds all current favorites
  const [favorites, setFavorites] = useState(mock);

  async function getData() {
    // GETS the newFavorites added to the favorites end-point from the search page
    try {
      const res = await Api.get("/favorites");
      setFavorites(res);
    } catch (error) {
      console.log(error);
    }
  }

  // This useEffect make sure that getData only fires once on render
  useEffect(() => {
    getData();
  }, []);

  return (
    <div className="container mx-auto">
      <div className="mt-10 text-white text-base">
        {/* Displays the Headers for the Favorite results */}
        <div className="mb-12">
          {favorites.length > 0 && (
            <div className="grid grid-cols-8 sticky top-0 bg-[rgb(22,22,22)]">
              <div className="p-3 col-span-1">#</div>
              <div className="p-3 col-span-3">Title</div>
              <div className="p-3 col-span-3">Collection</div>
              <div className="p-3 col-span-1">
                <HeartIcon className="w-7 h-7" />
              </div>
            </div>
          )}

          <div className="rounded-md flex flex-col gap-3">
            {/* This will loops through the current favorites so that it can be displayed */}
            {favorites.length === 0 && (
              <div className="text-3xl text-center">No current favorites</div>
            )}
            {favorites.map((result, index) => (
              <FavoriteItem
                key={result.id}
                index={index}
                result={result}
                refresh={getData}
              />
            ))}
          </div>
        </div>
      </div>
    </div>
  );
}

FavoritesPage.propTypes = {
  mock: PropTypes.array,
};

export default FavoritesPage;
