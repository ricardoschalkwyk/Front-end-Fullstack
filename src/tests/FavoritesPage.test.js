import React from "react";
import { screen, render } from "@testing-library/react";
import renderer from "react-test-renderer";

import "@testing-library/jest-dom";

import FavoritesPage from "../pages/FavoritesPage";

const mock = [
  {
    artistName: "Royal Sadness",
    trackName: "Ali Gatie",
    trackId: 1471588612,
    id: "54f13784-7afe-42c3-a530-5480d33cb44c",
  },
];

it("renders without crashing", () => {
  const div = document.createElement("div");
  render(<FavoritesPage />, div);
});

it("Matches Snapshot", () => {
  const tree = renderer.create(<FavoritesPage mock={mock} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Favorites must have mock data", () => {
  render(<FavoritesPage mock={mock} />);
  const element = screen.getByTestId("1");
  expect(element).toHaveTextContent("Ali Gatie");
});
