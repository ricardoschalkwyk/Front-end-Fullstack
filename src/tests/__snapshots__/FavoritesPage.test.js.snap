// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches Snapshot 1`] = `
<div
  className="container mx-auto"
>
  <div
    className="mt-10 text-white text-base"
  >
    <div
      className="mb-12"
    >
      <div
        className="grid grid-cols-8 sticky top-0 bg-[rgb(22,22,22)]"
      >
        <div
          className="p-3 col-span-1"
        >
          #
        </div>
        <div
          className="p-3 col-span-3"
        >
          Title
        </div>
        <div
          className="p-3 col-span-3"
        >
          Collection
        </div>
        <div
          className="p-3 col-span-1"
        >
          <svg
            aria-hidden="true"
            className="w-7 h-7"
            fill="none"
            stroke="currentColor"
            strokeWidth={2}
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
              strokeLinecap="round"
              strokeLinejoin="round"
            />
          </svg>
        </div>
      </div>
      <div
        className="rounded-md flex flex-col gap-3"
      >
        <div
          className="items-center grid grid-cols-8 gap-y-1 bg-slate-700 hover:bg-slate-500 transition-all duration-300 rounded-md "
        >
          <div
            className="px-3 py-2 col-span-1"
          >
            1
          </div>
          <div
            className="px-3 py-2 col-span-3"
            data-testid={1}
          >
            Ali Gatie
          </div>
          <div
            className="px-3 py-2 col-span-3"
          >
            Royal Sadness
          </div>
          <div
            className="px-3 py-2 col-span-1"
          >
            <button
              className="px-3 py-1 bg-slate-500 rounded hover:bg-neutral-900 text-white"
              onClick={[Function]}
              type="button"
            >
              <svg
                aria-hidden="true"
                className="w-5 h-5 text-red-500"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M3.172 5.172a4 4 0 015.656 0L10 6.343l1.172-1.171a4 4 0 115.656 5.656L10 17.657l-6.828-6.829a4 4 0 010-5.656z"
                  fillRule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
